src/mmds/src/token.rs

```diff
@@ -0,0 +1,562 @@
+        }
+
+        Ok(())
+    }
+
+    /// Validate the token time to live against bounds.
+    fn check_ttl(ttl_seconds: u32) -> bool {
+        MIN_TOKEN_TTL_SECONDS <= ttl_seconds && ttl_seconds <= MAX_TOKEN_TTL_SECONDS
+    }
+
+    /// Compute expiry time in seconds by adding the time to live provided
+    /// to the current time measured in seconds.
+    fn compute_expiry(ttl_seconds: u32) -> u64 {
+        // Get time elapsed since UNIX_EPOCH.
+        let now_as_duration = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("System time is before UNIX_EPOCH.");
+        // Convert elapsed time into seconds.
+        let ttl_duration = Duration::from_secs(u64::from(ttl_seconds));
+
+        // Compute expiry by adding ttl value to current time in seconds.
+        // This addition is safe because ttl is verified beforehand and
+        // can never be more than 6h.
+        let expiry_duration = now_as_duration.add(ttl_duration);
+
+        expiry_duration.as_secs()
+    }
+}
+
+/// Structure for token information.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+struct Token {
+    version: u16,
+    // Nonce or Initialization Vector.
+    iv: [u8; IV_LEN],
+    // Encrypted expire time.
+    payload: [u8; PAYLOAD_LEN],
+    // Tag returned after encryption.
+    tag: [u8; TAG_LEN],
+}
+
+impl Token {
+    /// Create a new token struct.
+    fn new(iv: [u8; IV_LEN], payload: [u8; PAYLOAD_LEN], tag: [u8; TAG_LEN]) -> Self {
+        Token {
+            version: 1,
+            iv,
+            payload,
+            tag,
+        }
+    }
+
+    /// Encode token structure into a string using base64 encoding.
+    fn base64_encode_token(&self) -> Result<String, Error> {
+        let token_bytes: Vec<u8> = DefaultOptions::new()
+            .with_fixint_encoding()
+            .allow_trailing_bytes()
+            .with_limit(DESERIALIZATION_BYTES_LIMIT)
+            .serialize(self)
+            .map_err(|_| Error::TokenGeneration)?;
+
+        // Encode token structure bytes into base64.
+        Ok(base64::encode_config(token_bytes, base64::STANDARD))
+    }
+
+    /// Decode token structure from base64 string.
+    fn base64_decode_token(encoded_token: &str) -> Result<Self, Error> {
+        let token_bytes = base64::decode_config(encoded_token, base64::STANDARD)
+            .map_err(|_| Error::ExpiryExtraction)?;
+
+        let token: Token = DefaultOptions::new()
+            .with_fixint_encoding()
+            .allow_trailing_bytes()
+            .with_limit(DESERIALIZATION_BYTES_LIMIT)
+            .deserialize(&token_bytes)
+            .map_err(|_| Error::ExpiryExtraction)?;
+        Ok(token)
+    }
+
+    /// Validate constant fields inside the token structure.
+    fn sanity_check(&self) -> bool {
+        self.version == 1
+    }
+}
+
+fn read_into_buf_from_entropy_pool(buf: &mut [u8]) -> Result<(), io::Error> {
+    let mut random_file = File::open(Path::new(RANDOMNESS_POOL))?;
```

Comment:
Would it make sense to define this `fd` as member of `TokenAuthority` instead of opening and closing it each time we read ?

Reply:
Good idea!